### PR TITLE
[#1241] Pull only consecutive transactions chain for consensus

### PR DESCRIPTION
**Pull request by TheMMaciek**

### CODE REVIEW

File: src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala [hunk 1]
```diff
@@ -54,25 +53,31 @@ class PendingTransactionsMemPool[F[_]: Concurrent](
...
         _.sortBy(_.map(-_.transaction.fee.getOrElse(0L)).sum).flatten
       )
+
+  private def takeConsecutiveTransactions(prevLastTxRef: LastTransactionRef, txs: List[TransactionCacheData]): List[TransactionCacheData] = {
+    txs.find(_.transaction.lastTxRef == prevLastTxRef) match {
+      case Some(tx) =>
+        tx +: takeConsecutiveTransactions(LastTransactionRef(tx.transaction.hash, tx.transaction.ordinal), txs diff List(tx))
+      case None => Nil
+    }
+  }
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala:80]: we need to be sure this recursion will not blow up. either by tailrec or making it iterative.

File: src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala [hunk 2]
```diff
@@ -151,6 +153,66 @@ class PendingTransactionsMemPoolTest extends FreeSpec with IdiomaticMockito with
...
+
+      //not sending tx4 to pending to break the chain
+      (memPool.put(tx1.hash, tx1) >>
+        memPool.put(tx2.hash, tx2) >>
+        memPool.put(tx3.hash, tx3) >>
+        memPool.put(tx5.hash, tx5)).unsafeRunSync
+
+      val result = memPool.pull(10).unsafeRunSync
+
+      result shouldBe List(tx1, tx2, tx3, tx5).some
```
Comment (marcinwadon) [inline src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala:174]: should it return `tx5` if `tx4` is missing and `tx5` references to `tx4` not `tx3`?

File: src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala [hunk 3]
```diff
@@ -151,6 +153,66 @@ class PendingTransactionsMemPoolTest extends FreeSpec with IdiomaticMockito with
...
+    }
+  }
+
+  "takeConsecutiveTransactions" - {
+    val memPool = PendingTransactionsMemPool[IO](txChainService, rl)
+    val takeConsecutiveTransactions = PrivateMethod[List[TransactionCacheData]]('takeConsecutiveTransactions)
```
Comment (marcinwadon) [inline src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala:180]: why you cannot test consecutive transactions using public interface of pending tx mem pool? you don't need to test this private method, we should rely only on the API/public interface.
just put transactions to mem pool and make assertions what you get from `pull` method

File: src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala (hunk 1 above)
Comment (kpudlik) [reply src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala:80]: Yeah, add `@tailrec` to function

Comment (TheMMaciek) [reply src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala:80]: Sure will do that.

File: src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala (hunk 2 above)
Comment (TheMMaciek) [reply src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala:174]: Yes because tx5 is a test transaction and previous logic allowed all the test transactions in. So I'm not changing that. Also there is an assumption that test transactions can be a part of the chain - we discussed that but didn't come to a conclusion that it's that or the other way around, so I remember I said that I will assume that they can be a part of a chain.

File: src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala (hunk 3 above)
Comment (TheMMaciek) [reply src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala:180]: I guess that's just me being used to going from the ground up. I think it is easier and more edge cases can be found when you are testing smaller bits of code. Then when moving upwards only testing what has not been tested yet or even repeating test cases in the bigger picture (didn't do that here - true). That's just the approach I followed so far to be sure that changes I introduce work. But I get your point of testing the public API only and I think it's valid. If someone wanted to reimplement functionality of takeConsecutiveTransactions with other function, or nest it in public function where it's being used then that person would also need to move the test cases to check the public function instead. On the other hand I think that testing is not only for asserting on public API functionality but also for the developer to check if he implemented the code the way he intended - by testing that code out of context like this private method here that can be achieved. Anyway... ;) I will move the cases to be tested through public API, but I can't assure that we will not have this conversation again as for some reason going from the ground up is more natural and intuitive and I do it involuntarily ;).

Comment (marcinwadon) [reply src/test/scala/org/constellation/storage/transactions/PendingTransactionsMemPoolTest.scala:180]: I work in the same way but try to think what you should expect at the end, not how to implement a single step. It’s like a declarative vs imperative - not how but what. And based on that you can create test cases upfront and work on the implementation then to fulfill those cases

File: src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala [hunk 4]
```diff
@@ -54,25 +55,32 @@ class PendingTransactionsMemPool[F[_]: Concurrent](
...
         _.sortBy(_.map(-_.transaction.fee.getOrElse(0L)).sum).flatten
       )
+
+  private def takeConsecutiveTransactions(lastAcceptedTxRef: LastTransactionRef, txs: List[TransactionCacheData]): List[TransactionCacheData] = {
+    @tailrec
+    def loop(acc: List[TransactionCacheData], prevTxRef: LastTransactionRef, txs: List[TransactionCacheData]): List[TransactionCacheData] =
+      txs.find(_.transaction.lastTxRef == prevTxRef) match {
+        case Some(tx) =>
+          loop(tx +: acc, LastTransactionRef(tx.transaction.hash, tx.transaction.ordinal), txs diff List(tx))
+        case None => acc.reverse //to preserve order of the chain
```
Comment (marcinwadon) [inline src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala:79]: so if you need to reverse the `acc` at the end, maybe just switch append/prepend in the loop?

Comment (TheMMaciek) [reply src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala:79]: I guess it's less heavy with prepend and one reverse at the end as opposed to append for every element. Prepend is constant time, append is linear.
